pbh-verifier/src/PBHVerifier.sol

```diff
@@ -0,0 +1,165 @@
+        bytes callData;
+        bytes32 accountGasLimits;
+        uint256 preVerificationGas;
+        bytes32 gasFees;
+        bytes paymasterAndData;
+        bytes signature;
+    }
+
+    struct PBHPayload {
+        uint256 root;
+        uint256 nullifierHash;
+        ExternalNullifier externalNullifier;
+        uint256[8] proof;
+    }
+
+    /**
+    * External Nullifier struct
+    * @param pbhNonce              - A nonce between 0 and numPbhPerMonth.
+    * @param month                 - An integer representing the current month.
+    * @param year                  - An integer representing the current year.
+    */
+    struct ExternalNullifier {
+        uint8 pbhNonce;
+        uint16 month;
+        uint8 year;
+    }
+
+    ///////////////////////////////////////////////////////////////////////////////
+    ///                                  Vars                                  ///
+    //////////////////////////////////////////////////////////////////////////////
+
+    /// @dev The World ID instance that will be used for verifying proofs
+    IWorldIDGroups internal immutable worldId;
+
+    /// @dev The World ID group ID (always 1)
+    uint256 internal immutable groupId = 1;
+    
+    /// @dev Make this configurable
+    uint256 internal immutable numPbhPerMonth = 30;
+
+    /// @dev Whether a nullifier hash has been used already. Used to guarantee an action is only performed once by a single person
+    mapping(uint256 => bool) internal nullifierHashes;
+    
+    ///////////////////////////////////////////////////////////////////////////////
+    ///                                  Functions                             ///
+    //////////////////////////////////////////////////////////////////////////////
+
+    /// @param _worldId The WorldID instance that will verify the proofs
+    constructor(
+        IWorldIDGroups _worldId
+    ) {
+        worldId = _worldId;
+    }
+    
+    function verifyExternalNullifier(ExternalNullifier memory externalNullifer) public view {
+        require(externalNullifer.year == BokkyPooBahsDateTimeLibrary.getYear(block.timestamp), InvalidExternalNullifierYear()); 
+        require(externalNullifer.month == BokkyPooBahsDateTimeLibrary.getMonth(block.timestamp), InvalidExternalNullifierMonth()); 
+        require(externalNullifer.pbhNonce <= numPbhPerMonth, InvalidPbhNonce()); 
+    }
+
+    /// @param userOp A packed user operation, used to generate the signal hash
+    /// @param root The root of the Merkle tree (returned by the JS widget).
+    /// @param nullifierHash The nullifier hash for this proof, preventing double signaling (returned by the JS widget).
+    /// @param proof The zero-knowledge proof that demonstrates the claimer is registered with World ID (returned by the JS widget).
+    function verifyPbhProof(
+        PackedUserOperation memory userOp,
+        uint256 root,
+        uint256 nullifierHash,
+        ExternalNullifier memory externalNullifier,
```

Comment:
how are we getting the externalNullifier data here - do we need to add this with the proof in the signature

or is this determined by the builder?

Reply:
@karankurbur Yes, this should be decoded from the `signature` on the `UserOperation`